Modules/grpmodule.c

```diff
@@ -154,12 +186,42 @@ grp_getgrnam_impl(PyObject *module, PyObject *name)
     /* check for embedded null bytes */
     if (PyBytes_AsStringAndSize(bytes, &name_chars, NULL) == -1)
         goto out;
+#ifdef HAVE_GETGRNAM_R
+    Py_BEGIN_ALLOW_THREADS
+    int status, bufsize = NSS_BUFLEN_GROUP;
+    struct group grp;
+
+    buf = PyMem_RawMalloc(bufsize);
+
+    do {
+        status = getgrnam_r(name_chars, &grp, buf, bufsize, &p);
+        if (p != NULL || status != ERANGE) {
+            break;
+        }
+        if ((bufsize << 1) > (NSS_BUFLEN_GROUP << 10)) {
+            status = ERANGE;
+            break;
+        }
+        bufsize <<= 1;
+        buf = PyMem_RawRealloc(buf, bufsize);
+    } while (1);
 
-    if ((p = getgrnam(name_chars)) == NULL) {
+    if (status != 0 || p == NULL) {
+        PyMem_RawFree(buf);
+        p = NULL;
+    }
+    Py_END_ALLOW_THREADS
+#else
+    p = getgrnam(name_chars);
+#endif
+    if (p == NULL) {
         PyErr_Format(PyExc_KeyError, "getgrnam(): name not found: %s", name_chars);
         goto out;
```

Comment:
You leak buf here.

Reply:
I am failing to see how. If p == NULL it will free buf in the if statement before.